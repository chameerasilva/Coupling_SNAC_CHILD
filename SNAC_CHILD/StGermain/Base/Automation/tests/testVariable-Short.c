/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
**
** Copyright (C), 2003, Victorian Partnership for Advanced Computing (VPAC) Ltd, 110 Victoria Street, Melbourne, 3053, Australia.
**
** Authors:
**	Stevan M. Quenette, Senior Software Engineer, VPAC. (steve@vpac.org)
**	Patrick D. Sunter, Software Engineer, VPAC. (pds@vpac.org)
**	Luke J. Hodkinson, Computational Engineer, VPAC. (lhodkins@vpac.org)
**	Siew-Ching Tan, Software Engineer, VPAC. (siew@vpac.org)
**	Alan H. Lo, Computational Engineer, VPAC. (alan@vpac.org)
**	Raquibul Hassan, Computational Engineer, VPAC. (raq@vpac.org)
**
**  This library is free software; you can redistribute it and/or
**  modify it under the terms of the GNU Lesser General Public
**  License as published by the Free Software Foundation; either
**  version 2.1 of the License, or (at your option) any later version.
**
**  This library is distributed in the hope that it will be useful,
**  but WITHOUT ANY WARRANTY; without even the implied warranty of
**  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
**  Lesser General Public License for more details.
**
**  You should have received a copy of the GNU Lesser General Public
**  License along with this library; if not, write to the Free Software
**  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
**
** Role:
** 	Tests the Short macros of Variable. Will test a single scalar, vector and each scalar created from the vector.
**
** $Id: testVariable-Short.c 2136 2004-09-30 02:47:13Z PatrickSunter $
**
**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

#include <mpi.h>
#include "Base/Foundation/Foundation.h"
#include "Base/IO/IO.h"
#include "Base/Container/Container.h"
#include "Base/Automation/Automation.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>


int main( int argc, char *argv[] ) {
	int		rank;
	int		procCount;
	int		procToWatch;
	Stream*		stream;
	
	/* Initialise MPI, get world info */
	MPI_Init( &argc, &argv );
	MPI_Comm_size( MPI_COMM_WORLD, &procCount );
	MPI_Comm_rank( MPI_COMM_WORLD, &rank );
	
	BaseFoundation_Init( &argc, &argv );
	BaseIO_Init( &argc, &argv );
	BaseContainer_Init( &argc, &argv );
	BaseAutomation_Init( &argc, &argv );

	stream = Journal_Register( Info_Type, "testVariable" );
	
	if( argc >= 2 ) {
		procToWatch = atoi( argv[1] );
	}
	else {
		procToWatch = 0;
	}
	if( rank == procToWatch ) {
		typedef short Triple[3];

		short* array;
		Triple* structArray;
		Index length = 10;

		/* List of values to test the variable with.
		 * Values to test are hex 5's and a's because they are a series of 0101 and 1010 respectively so they test
		 * each bit in memory to read/set.
		 */
		long int testValues[] = { 0x5555, 0xaaaa };
		Index testValueCount = 2;
		Index test_I;
		long int testValue;

		Variable* var;
		Variable* vec;
		Variable* vecVar[3];

		Variable_Register* reg;

		int i, j;

		array = Memory_Alloc_Array( short, length, "test" );
		structArray = Memory_Alloc_Array( Triple, length, "test" );

		reg = Variable_Register_New();

		var = Variable_NewScalar(
			"Scalar",
			Variable_DataType_Short,
			&length,
			(void**)&array,
			reg );

		vec = Variable_NewVector(
			"Three",
			Variable_DataType_Short,
			3,
			&length,
			(void**)&structArray,
			reg,
			"a",
			"b",
			"c" );
		vecVar[0] = Variable_Register_GetByName( reg, "a" );
		vecVar[1] = Variable_Register_GetByName( reg, "b" );
		vecVar[2] = Variable_Register_GetByName( reg, "c" );

		Variable_Register_BuildAll( reg );

		Print( reg, stream );

		for ( test_I = 0; test_I < testValueCount; ++test_I ) {	

			testValue = testValues[test_I];

			Journal_Printf( stream, "Setting value %d\n", (short)(short)testValue );

			for ( i = 0; i < length; ++i ) {
				Variable_SetValueShort( var, i, testValue );

				Variable_SetValueAtShort( vec, i, 0, testValue );
				Variable_SetValueAtShort( vec, i, 1, testValue );
				Variable_SetValueAtShort( vec, i, 2, testValue );
			}

			Journal_Printf( stream, "~~~Scalar~~~\n" );
			for ( i = 0; i < length; ++i ) {
				Journal_Printf( stream, "i=%d: %d %d %d %d %d \n",
						i,
						Variable_GetValueShort( var, i ) == (short)(short)testValue,
						Variable_GetValueShortAsChar( var, i ) == (char)(short)testValue,
						Variable_GetValueShortAsInt( var, i ) == (int)(short)testValue,
						Variable_GetValueShortAsFloat( var, i ) == (float)(short)testValue,
						Variable_GetValueShortAsDouble( var, i ) == (double)(short)testValue );
			}

			Journal_Printf( stream, "~~~Vector~~~\n" );
			for ( i = 0; i < length; ++i ) {
				Journal_Printf( stream, "a, i=%d: %d %d %d %d %d ",
						i,
						Variable_GetValueAtShort( vec, i, 0 ) == (short)(short)testValue,
						Variable_GetValueAtShortAsChar( vec, i, 0 ) == (char)(short)testValue,
						Variable_GetValueAtShortAsInt( vec, i, 0 ) == (int)(short)testValue,
						Variable_GetValueAtShortAsFloat( vec, i, 0 ) == (float)(short)testValue,
						Variable_GetValueAtShortAsDouble( vec, i, 0 ) == (double)(short)testValue );
				Journal_Printf( stream, "ptr test %s\n",
						( Variable_GetPtrAtShort( vec, i, 0 ) == &structArray[i][0] ) ? "OK" : "FAILED" );

				Journal_Printf( stream, "b, i=%d: %d %d %d %d %d ",
						i,
						Variable_GetValueAtShort( vec, i, 1 ) == (short)(short)testValue,
						Variable_GetValueAtShortAsChar( vec, i, 1 ) == (char)(short)testValue,
						Variable_GetValueAtShortAsInt( vec, i, 1 ) == (int)(short)testValue,
						Variable_GetValueAtShortAsFloat( vec, i, 1 ) == (float)(short)testValue,
						Variable_GetValueAtShortAsDouble( vec, i, 1 ) == (double)(short)testValue );
				Journal_Printf( stream, "ptr test %s\n",
						( Variable_GetPtrAtShort( vec, i, 1 ) == &structArray[i][1] ) ? "OK" : "FAILED" );

				Journal_Printf( stream, "c, i=%d: %d %d %d %d %d ",
						i,
						Variable_GetValueAtShort( vec, i, 2 ) == (short)(short)testValue,
						Variable_GetValueAtShortAsChar( vec, i, 2 ) == (char)(short)testValue,
						Variable_GetValueAtShortAsInt( vec, i, 2 ) == (int)(short)testValue,
						Variable_GetValueAtShortAsFloat( vec, i, 2 ) == (float)(short)testValue,
						Variable_GetValueAtShortAsDouble( vec, i, 2 ) == (double)(short)testValue );
				Journal_Printf( stream, "ptr test %s\n",
						( Variable_GetPtrAtShort( vec, i, 2 ) == &structArray[i][2] ) ? "OK" : "FAILED" );

			}

			Journal_Printf( stream, "~~~Vector: Sub-Variable~~~\n" );
			for ( i = 0; i < length; ++i ) {
				for ( j = 0; j < 3; ++j ) {
					Journal_Printf( stream, "j=%d, i=%d: %s\n",
						j,
						i,
						( _Variable_GetStructPtr( vecVar[j], i ) == &structArray[i][j] ) ? "OK" : "FAIL" );
				}
			}

			
		}
	
	}
	
	BaseAutomation_Finalise();
	BaseContainer_Finalise();
	BaseIO_Finalise();
	BaseFoundation_Finalise();

	
	/* Close off MPI */
	MPI_Finalize();
	
	return 0; /* success */
}
