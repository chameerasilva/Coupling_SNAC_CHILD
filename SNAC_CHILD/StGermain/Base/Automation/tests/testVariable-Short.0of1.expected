Variable_Register (ptr): 0x805a7a0
	count: 5
	_size: 8
	_delta: 8
	_variable (ptr): 0x805ae50
Stg_Component (ptr): 0x805ae80
Stg_Object (ptr): 0x805ae80
	Stg_Class (ptr): 0x805ae80
		sizeOfSelf: 152
		_deleteSelf: Yes
		type: Variable
		_delete (func ptr): 0xb7f3e4a0
		_print (func ptr): 0xb7f3e538
		_copy (func ptr): 0xb7f3e921
	name: Scalar
	nameAllocationType: NON_GLOBAL
	_defaultConstructor (func ptr): 0xb7f3d986
	_construct (func ptr): 0xb7f3f5c2
	_build (func ptr): 0xb7f3f0a2
	_initialise (func ptr): 0xb7f3f5b8
	_execute (func ptr): 0xb7f3f5bd
	_destroy (func ptr): 0xb7f3fb1a
	isConstructed: True
	isBuilt: True
	isInitialised: False
	hasExecuted: False
	isDestroyed: False
	constructor function name: Scalar-Construct
	build function name: Scalar-Build
	initialise function name: Scalar-Initialise
	execute function name: Scalar-Execute
	destroy function name: Scalar-Destroy
Variable (ptr): (0x805ae80)
	offsetCount: 1
	offsets (ptr): 0x805acd0
		offsets[0]: 0
	dataTypes (ptr): 0x805ac48
		dataTypes[0]: 2
	dataTypeCounts (ptr): 0x805b338
		dataTypeCounts[0]: 1
	dataSizes (ptr): 0x805c030
		dataSizes[0]: 2
	components (ptr): 0x805b348
		components[0] (ptr): (nil)
	structSize: 2
	structSizePtr (ptr): (nil)
	arrayPtr (ptr): 0x805ade8
	arrayPtrPtr (ptr): 0xbfd6753c
	arraySize: 10
	arraySizePtr (ptr): 0xbfd67534
	allocateSelf = False
Stg_Component (ptr): 0x805b358
Stg_Object (ptr): 0x805b358
	Stg_Class (ptr): 0x805b358
		sizeOfSelf: 152
		_deleteSelf: Yes
		type: Variable
		_delete (func ptr): 0xb7f3e4a0
		_print (func ptr): 0xb7f3e538
		_copy (func ptr): 0xb7f3e921
	name: Three
	nameAllocationType: NON_GLOBAL
	_defaultConstructor (func ptr): 0xb7f3d986
	_construct (func ptr): 0xb7f3f5c2
	_build (func ptr): 0xb7f3f0a2
	_initialise (func ptr): 0xb7f3f5b8
	_execute (func ptr): 0xb7f3f5bd
	_destroy (func ptr): 0xb7f3fb1a
	isConstructed: True
	isBuilt: True
	isInitialised: False
	hasExecuted: False
	isDestroyed: False
	constructor function name: Three-Construct
	build function name: Three-Build
	initialise function name: Three-Initialise
	execute function name: Three-Execute
	destroy function name: Three-Destroy
Variable (ptr): (0x805b358)
	offsetCount: 1
	offsets (ptr): 0x805b648
		offsets[0]: 0
	dataTypes (ptr): 0x805b658
		dataTypes[0]: 2
	dataTypeCounts (ptr): 0x805b668
		dataTypeCounts[0]: 3
	dataSizes (ptr): 0x805c040
		dataSizes[0]: 6
	components (ptr): 0x805b678
		components[0] (ptr): 0x805b690
		components[0]: a
	structSize: 6
	structSizePtr (ptr): (nil)
	arrayPtr (ptr): 0x805ae08
	arrayPtrPtr (ptr): 0xbfd67538
	arraySize: 10
	arraySizePtr (ptr): 0xbfd67534
	allocateSelf = False
Stg_Component (ptr): 0x805b690
Stg_Object (ptr): 0x805b690
	Stg_Class (ptr): 0x805b690
		sizeOfSelf: 152
		_deleteSelf: Yes
		type: Variable
		_delete (func ptr): 0xb7f3e4a0
		_print (func ptr): 0xb7f3e538
		_copy (func ptr): 0xb7f3e921
	name: a
	nameAllocationType: NON_GLOBAL
	_defaultConstructor (func ptr): 0xb7f3d986
	_construct (func ptr): 0xb7f3f5c2
	_build (func ptr): 0xb7f3f0a2
	_initialise (func ptr): 0xb7f3f5b8
	_execute (func ptr): 0xb7f3f5bd
	_destroy (func ptr): 0xb7f3fb1a
	isConstructed: True
	isBuilt: True
	isInitialised: False
	hasExecuted: False
	isDestroyed: False
	constructor function name: a-Construct
	build function name: a-Build
	initialise function name: a-Initialise
	execute function name: a-Execute
	destroy function name: a-Destroy
Variable (ptr): (0x805b690)
	offsetCount: 1
	offsets (ptr): 0x805b978
		offsets[0]: 0
	dataTypes (ptr): 0x805b988
		dataTypes[0]: 2
	dataTypeCounts (ptr): 0x805b998
		dataTypeCounts[0]: 1
	dataSizes (ptr): 0x805c050
		dataSizes[0]: 2
	components (ptr): 0x805b9a8
		components[0] (ptr): (nil)
	structSize: 6
	structSizePtr (ptr): (nil)
	arrayPtr (ptr): 0x805ae08
	arrayPtrPtr (ptr): 0xbfd67538
	arraySize: 10
	arraySizePtr (ptr): 0xbfd67534
	allocateSelf = False
Stg_Component (ptr): 0x805b9b8
Stg_Object (ptr): 0x805b9b8
	Stg_Class (ptr): 0x805b9b8
		sizeOfSelf: 152
		_deleteSelf: Yes
		type: Variable
		_delete (func ptr): 0xb7f3e4a0
		_print (func ptr): 0xb7f3e538
		_copy (func ptr): 0xb7f3e921
	name: b
	nameAllocationType: NON_GLOBAL
	_defaultConstructor (func ptr): 0xb7f3d986
	_construct (func ptr): 0xb7f3f5c2
	_build (func ptr): 0xb7f3f0a2
	_initialise (func ptr): 0xb7f3f5b8
	_execute (func ptr): 0xb7f3f5bd
	_destroy (func ptr): 0xb7f3fb1a
	isConstructed: True
	isBuilt: True
	isInitialised: False
	hasExecuted: False
	isDestroyed: False
	constructor function name: b-Construct
	build function name: b-Build
	initialise function name: b-Initialise
	execute function name: b-Execute
	destroy function name: b-Destroy
Variable (ptr): (0x805b9b8)
	offsetCount: 1
	offsets (ptr): 0x805bca0
		offsets[0]: 2
	dataTypes (ptr): 0x805bcb0
		dataTypes[0]: 2
	dataTypeCounts (ptr): 0x805bcc0
		dataTypeCounts[0]: 1
	dataSizes (ptr): 0x805c060
		dataSizes[0]: 2
	components (ptr): 0x805bcd0
		components[0] (ptr): (nil)
	structSize: 6
	structSizePtr (ptr): (nil)
	arrayPtr (ptr): 0x805ae08
	arrayPtrPtr (ptr): 0xbfd67538
	arraySize: 10
	arraySizePtr (ptr): 0xbfd67534
	allocateSelf = False
Stg_Component (ptr): 0x805bce0
Stg_Object (ptr): 0x805bce0
	Stg_Class (ptr): 0x805bce0
		sizeOfSelf: 152
		_deleteSelf: Yes
		type: Variable
		_delete (func ptr): 0xb7f3e4a0
		_print (func ptr): 0xb7f3e538
		_copy (func ptr): 0xb7f3e921
	name: c
	nameAllocationType: NON_GLOBAL
	_defaultConstructor (func ptr): 0xb7f3d986
	_construct (func ptr): 0xb7f3f5c2
	_build (func ptr): 0xb7f3f0a2
	_initialise (func ptr): 0xb7f3f5b8
	_execute (func ptr): 0xb7f3f5bd
	_destroy (func ptr): 0xb7f3fb1a
	isConstructed: True
	isBuilt: True
	isInitialised: False
	hasExecuted: False
	isDestroyed: False
	constructor function name: c-Construct
	build function name: c-Build
	initialise function name: c-Initialise
	execute function name: c-Execute
	destroy function name: c-Destroy
Variable (ptr): (0x805bce0)
	offsetCount: 1
	offsets (ptr): 0x805bfc8
		offsets[0]: 4
	dataTypes (ptr): 0x805bfd8
		dataTypes[0]: 2
	dataTypeCounts (ptr): 0x805bfe8
		dataTypeCounts[0]: 1
	dataSizes (ptr): 0x805c070
		dataSizes[0]: 2
	components (ptr): 0x805bff8
		components[0] (ptr): (nil)
	structSize: 6
	structSizePtr (ptr): (nil)
	arrayPtr (ptr): 0x805ae08
	arrayPtrPtr (ptr): 0xbfd67538
	arraySize: 10
	arraySizePtr (ptr): 0xbfd67534
	allocateSelf = False
Stg_Class (ptr): 0x805a7a0
	sizeOfSelf: 40
	_deleteSelf: Yes
	type: Variable_Register
	_delete (func ptr): 0xb7f40561
	_print (func ptr): 0xb7f405a1
	_copy (func ptr): 0xb7f406ca
Setting value 21845
~~~Scalar~~~
i=0: 1 1 1 1 1 
i=1: 1 1 1 1 1 
i=2: 1 1 1 1 1 
i=3: 1 1 1 1 1 
i=4: 1 1 1 1 1 
i=5: 1 1 1 1 1 
i=6: 1 1 1 1 1 
i=7: 1 1 1 1 1 
i=8: 1 1 1 1 1 
i=9: 1 1 1 1 1 
~~~Vector~~~
a, i=0: 1 1 1 1 1 ptr test OK
b, i=0: 1 1 1 1 1 ptr test OK
c, i=0: 1 1 1 1 1 ptr test OK
a, i=1: 1 1 1 1 1 ptr test OK
b, i=1: 1 1 1 1 1 ptr test OK
c, i=1: 1 1 1 1 1 ptr test OK
a, i=2: 1 1 1 1 1 ptr test OK
b, i=2: 1 1 1 1 1 ptr test OK
c, i=2: 1 1 1 1 1 ptr test OK
a, i=3: 1 1 1 1 1 ptr test OK
b, i=3: 1 1 1 1 1 ptr test OK
c, i=3: 1 1 1 1 1 ptr test OK
a, i=4: 1 1 1 1 1 ptr test OK
b, i=4: 1 1 1 1 1 ptr test OK
c, i=4: 1 1 1 1 1 ptr test OK
a, i=5: 1 1 1 1 1 ptr test OK
b, i=5: 1 1 1 1 1 ptr test OK
c, i=5: 1 1 1 1 1 ptr test OK
a, i=6: 1 1 1 1 1 ptr test OK
b, i=6: 1 1 1 1 1 ptr test OK
c, i=6: 1 1 1 1 1 ptr test OK
a, i=7: 1 1 1 1 1 ptr test OK
b, i=7: 1 1 1 1 1 ptr test OK
c, i=7: 1 1 1 1 1 ptr test OK
a, i=8: 1 1 1 1 1 ptr test OK
b, i=8: 1 1 1 1 1 ptr test OK
c, i=8: 1 1 1 1 1 ptr test OK
a, i=9: 1 1 1 1 1 ptr test OK
b, i=9: 1 1 1 1 1 ptr test OK
c, i=9: 1 1 1 1 1 ptr test OK
~~~Vector: Sub-Variable~~~
j=0, i=0: OK
j=1, i=0: OK
j=2, i=0: OK
j=0, i=1: OK
j=1, i=1: OK
j=2, i=1: OK
j=0, i=2: OK
j=1, i=2: OK
j=2, i=2: OK
j=0, i=3: OK
j=1, i=3: OK
j=2, i=3: OK
j=0, i=4: OK
j=1, i=4: OK
j=2, i=4: OK
j=0, i=5: OK
j=1, i=5: OK
j=2, i=5: OK
j=0, i=6: OK
j=1, i=6: OK
j=2, i=6: OK
j=0, i=7: OK
j=1, i=7: OK
j=2, i=7: OK
j=0, i=8: OK
j=1, i=8: OK
j=2, i=8: OK
j=0, i=9: OK
j=1, i=9: OK
j=2, i=9: OK
Setting value -21846
~~~Scalar~~~
i=0: 1 1 1 1 1 
i=1: 1 1 1 1 1 
i=2: 1 1 1 1 1 
i=3: 1 1 1 1 1 
i=4: 1 1 1 1 1 
i=5: 1 1 1 1 1 
i=6: 1 1 1 1 1 
i=7: 1 1 1 1 1 
i=8: 1 1 1 1 1 
i=9: 1 1 1 1 1 
~~~Vector~~~
a, i=0: 1 1 1 1 1 ptr test OK
b, i=0: 1 1 1 1 1 ptr test OK
c, i=0: 1 1 1 1 1 ptr test OK
a, i=1: 1 1 1 1 1 ptr test OK
b, i=1: 1 1 1 1 1 ptr test OK
c, i=1: 1 1 1 1 1 ptr test OK
a, i=2: 1 1 1 1 1 ptr test OK
b, i=2: 1 1 1 1 1 ptr test OK
c, i=2: 1 1 1 1 1 ptr test OK
a, i=3: 1 1 1 1 1 ptr test OK
b, i=3: 1 1 1 1 1 ptr test OK
c, i=3: 1 1 1 1 1 ptr test OK
a, i=4: 1 1 1 1 1 ptr test OK
b, i=4: 1 1 1 1 1 ptr test OK
c, i=4: 1 1 1 1 1 ptr test OK
a, i=5: 1 1 1 1 1 ptr test OK
b, i=5: 1 1 1 1 1 ptr test OK
c, i=5: 1 1 1 1 1 ptr test OK
a, i=6: 1 1 1 1 1 ptr test OK
b, i=6: 1 1 1 1 1 ptr test OK
c, i=6: 1 1 1 1 1 ptr test OK
a, i=7: 1 1 1 1 1 ptr test OK
b, i=7: 1 1 1 1 1 ptr test OK
c, i=7: 1 1 1 1 1 ptr test OK
a, i=8: 1 1 1 1 1 ptr test OK
b, i=8: 1 1 1 1 1 ptr test OK
c, i=8: 1 1 1 1 1 ptr test OK
a, i=9: 1 1 1 1 1 ptr test OK
b, i=9: 1 1 1 1 1 ptr test OK
c, i=9: 1 1 1 1 1 ptr test OK
~~~Vector: Sub-Variable~~~
j=0, i=0: OK
j=1, i=0: OK
j=2, i=0: OK
j=0, i=1: OK
j=1, i=1: OK
j=2, i=1: OK
j=0, i=2: OK
j=1, i=2: OK
j=2, i=2: OK
j=0, i=3: OK
j=1, i=3: OK
j=2, i=3: OK
j=0, i=4: OK
j=1, i=4: OK
j=2, i=4: OK
j=0, i=5: OK
j=1, i=5: OK
j=2, i=5: OK
j=0, i=6: OK
j=1, i=6: OK
j=2, i=6: OK
j=0, i=7: OK
j=1, i=7: OK
j=2, i=7: OK
j=0, i=8: OK
j=1, i=8: OK
j=2, i=8: OK
j=0, i=9: OK
j=1, i=9: OK
j=2, i=9: OK
